File: StudyOC/StudyOC/CGD定时器/TimerController.m
//
//  TimerController.m
//  StudyOC
//
//  Created by z Hahn on 2022/6/20.
//

#import "TimerController.h"
#import "GCDTimer.h"


@interface TimerController ()

@property (strong, nonatomic) GCDTimer * timer;

@end

@implementation TimerController


- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    NSLog(@"name = %@",self);
}

- (void)viewDidDisappear:(BOOL)animated
{
    [super viewDidDisappear:animated];
    [self.timer invalidate];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.timer = [GCDTimer scheduledTimerWithTimeInterval:1 repeat:YES block:^{
        NSLog(@"执行");
        
        [self.timer stop];
        [self.timer restart];
    }];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: StudySwift/StudySwift/CGD定时器/TimerController.swift
//
//  TimerController.swift
//  体系_Swift
//
//  Created by z Hahn on 2022/6/20.
//

import UIKit

class TimerController: UIViewController {
    
    private var timer: CGDTimer = CGDTimer()
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        print("name = \(self)")
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        CGDTimer.default.invalidate()
        self.timer.invalidate()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // 单例创建主线程定时器
//        CGDTimer.default.setTimeInterval(interval: 1, isRepeat: true) {
//            print("执行返回")
//
//            CGDTimer.default.stop()
//
//            CGDTimer.default.restart()
//        }
        
        // 单例创建自定义程定时器
//        CGDTimer.default.setTimeInterval(ztimer: DispatchSource.makeTimerSource(), interval: 1, isRepeat: true) {
//            print("执行返回")
//
//            CGDTimer.default.stop()
//
//            CGDTimer.default.restart()
//        }
        
        // 对象创建主线程定时器
//        timer.setTimeInterval(interval: 1, isRepeat: true) { [self] in
//            print("执行返回")
//            timer.stop()
//            timer.restart()
//        }
        
        // 对象创建自定义线程定时器
        timer.setTimeInterval(ztimer: DispatchSource.makeTimerSource(), interval: 1, isRepeat: true) { [self] in
            print("执行返回")
            timer.stop()
            timer.restart()
        }
    }
    


}
